borsh-derive-internal/src/attribute_helpers.rs

```diff
@@ -42,6 +46,80 @@ pub(crate) fn contains_skip(attrs: &[Attribute]) -> bool {
     attrs.iter().any(|attr| attr.path() == SKIP)
 }
 
+pub fn check_item_attributes(derive_input: &DeriveInput) -> Result<(), TokenStream> {
+    for attr in &derive_input.attrs {
+        if attr.path().is_ident("borsh") {
+            attr.parse_nested_meta(|meta| {
+                if !meta.path.is_ident(USE_DISCRIMINANT) {
+                    return Err(syn::Error::new(
+                        derive_input.ident.span(),
+                        "`use_discriminant` is the only supported attribute for `borsh`",
+                    ));
+                }
+                if meta.path.is_ident(USE_DISCRIMINANT) {
+                    let _expr: Expr = meta.value()?.parse()?;
+                    if let syn::Data::Struct(ref _data) = derive_input.data {
+                        return Err(syn::Error::new(
+                            derive_input.ident.span(),
+                            "borsh(use_discriminant=<bool>) does not support structs",
+                        ));
+                    }
+                }
+
+                Ok(())
+            })
+            .map_err(|err| err.to_compile_error())?;
+        }
+    }
+    Ok(())
+}
+
+pub(crate) fn contains_use_discriminant(input: &ItemEnum) -> Result<bool, syn::Error> {
+    assert!(
+        input.variants.len() < 256,
+        "up to 256 enum variants are supported"
+    );
```

Comment:
replace panic with returning `syn::Error::new` with span on `input` .
I don't see a reason why use panic here instead of returning Error, as in other cases